File: src/update_student_ids.js
import fs from 'fs';
import csv from 'csv-parser';
import { createObjectCsvWriter } from 'csv-writer';

const STUDENT_EXIST_PATH = '/Users/htoomaung/Repository/withus-kodai-admission-migration/src/student-exist.csv';
const BASE_FILE_PATH = '/Users/htoomaung/Repository/withus-kodai-admission-migration/src/processed_student_data_fixed_masked.csv';

// First, read student-exist.csv to create the mapping
console.log('Reading student-exist.csv...');
const idMapping = new Map();

fs.createReadStream(STUDENT_EXIST_PATH)
    .pipe(csv())
    .on('data', (row) => {
        idMapping.set(row['MANAERP__External_User_Id__c'], row['Id']);
    })
    .on('end', () => {
        console.log('Processing base file...');
        
        // Get headers from the input file
        const headers = [];
        fs.createReadStream(BASE_FILE_PATH)
            .pipe(csv())
            .on('data', (row) => {
                if (headers.length === 0) {
                    headers.push(...Object.keys(row));
                    // Add Id column if it doesn't exist
                    if (!headers.includes('Id')) {
                        headers.push('Id');
                    }
                }
            })
            .on('end', () => {
                // Create CSV writer with the same headers and always quote fields
                const csvWriter = createObjectCsvWriter({
                    path: BASE_FILE_PATH,
                    header: headers.map(header => ({ id: header, title: header })),
                    alwaysQuote: true
                });

                const records = [];
                
                // Process the input file
                fs.createReadStream(BASE_FILE_PATH)
                    .pipe(csv())
                    .on('data', (row) => {
                        // Create a new object with all values preserved as strings
                        const newRow = {};
                        for (const key in row) {
                            newRow[key] = row[key];
                        }
                        // Add Id field if it doesn't exist
                        if (!('Id' in newRow)) {
                            newRow['Id'] = '';
                        }
                        // Update the Id field
                        newRow['Id'] = idMapping.get(row['MANAERP__External_User_Id__c']) || '';
                        records.push(newRow);
                    })
                    .on('end', () => {
                        // Write all records back to the base file
                        csvWriter.writeRecords(records)
                            .then(() => {
                                console.log('Processing complete! Updated Id values in the base file.');
                            });
                    });
            });
    }); 

File: src/update_student_ids.py
import pandas as pd

# Read the CSV files with string dtype to prevent any conversion
print("Reading student-exist.csv...")
student_exist_df = pd.read_csv('/Users/htoomaung/Repository/withus-kodai-admission-migration/src/student-exist.csv', 
                             dtype=str,
                             keep_default_na=False)

# Create a mapping of External User ID to Id
id_mapping = dict(zip(student_exist_df['MANAERP__External_User_Id__c'], student_exist_df['Id']))

# Read the base file in chunks to handle large file
print("Processing base file...")
chunk_size = 10000
output_file = '/Users/htoomaung/Repository/withus-kodai-admission-migration/src/processed_student_data_updated.csv'

# Process the file in chunks, reading all as strings
for chunk in pd.read_csv('/Users/htoomaung/Repository/withus-kodai-admission-migration/src/processed_student_data_fixed_masked.csv', 
                        chunksize=chunk_size,
                        dtype=str,
                        keep_default_na=False):
    # Update the Id column where External User ID matches
    chunk['Id'] = chunk['MANAERP__External_User_Id__c'].map(id_mapping)
    
    # Write the chunk to the output file, ensuring all values are written as strings
    chunk.to_csv(output_file, 
                mode='a', 
                header=not pd.io.common.file_exists(output_file), 
                index=False,
                quoting=1)  # Quote all fields to preserve exact string values

print("Processing complete! Updated file saved as processed_student_data_updated.csv") 